app-libs/sgx-runtime/pallets/guess-the-number/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+
+	#[pallet::storage]
+	#[pallet::getter(fn lucky_number)]
+	pub(super) type LuckyNumber<T: Config> = StorageValue<_, GuessType, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn guess_attempts)]
+	pub(super) type GuessAttempts<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, u8, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn last_winning_distance)]
+	pub(super) type LastWinningDistance<T: Config> = StorageValue<_, GuessType, OptionQuery>;
+	#[pallet::storage]
+	#[pallet::getter(fn winning_distance)]
+	pub(super) type WinningDistance<T: Config> = StorageValue<_, GuessType, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn last_winners)]
+	pub(super) type LastWinners<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn winners)]
+	pub(super) type Winners<T: Config> = StorageValue<_, Vec<T::AccountId>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn winnings)]
+	pub(super) type Winnings<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::type_value]
+	pub(super) fn DefaultForNextRoundTimestamp<T: Config>() -> T::Moment {
+		T::Moment::zero()
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn next_round_timestamp)]
+	pub(super) type NextRoundTimestamp<T: Config> =
+		StorageValue<_, T::Moment, ValueQuery, DefaultForNextRoundTimestamp<T>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+	{
+		/// Push next round by one entire day
+		///
+		/// May only be called from `T::GameMaster`.
+		#[pallet::call_index(0)]
+		#[pallet::weight((<T as Config>::WeightInfo::push_by_one_day(), DispatchClass::Normal, Pays::Yes)
+        )]
+		pub fn push_by_one_day(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			T::GameMaster::ensure_origin(origin)?;
+			let tnext = Self::next_round_timestamp().saturating_add(T::MomentsPerDay::get());
+			<NextRoundTimestamp<T>>::put(tnext);
+			debug!("pushing next round by one day");
+			Self::deposit_event(Event::RoundSchedulePushedByOneDay);
+			Ok(().into())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight((<T as Config>::WeightInfo::set_winnings(), DispatchClass::Normal, Pays::Yes)
+        )]
+		pub fn set_winnings(
+			origin: OriginFor<T>,
+			winnings: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			T::GameMaster::ensure_origin(origin)?;
+			debug!("setting winnings");
+			<Winnings<T>>::put(winnings);
+			Ok(().into())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight((<T as Config>::WeightInfo::guess(), DispatchClass::Normal, Pays::Yes)
+        )]
+		pub fn guess(origin: OriginFor<T>, guess: GuessType) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(
+				Self::guess_attempts(&sender) < T::MaxAttempts::get(),
+				Error::<T>::TooManyAttempts
+			);
+			<GuessAttempts<T>>::mutate(&sender, |a| *a += 1u8);
```

Comment:
I guess we don't need the saturated add if we ensure < T::MaxAttempts right above.

Reply:
hmm, better to do it explicitly....